pkg/query/flamegraph_arrow.go

```diff
@@ -832,99 +856,133 @@ func newFlamegraphBuilder(
 	return fb, nil
 }
 
-// NewRecord returns a new record from the builders.
-// It adds the children to the children column and the labels intersection to the labels column.
-// Finally, it assembles all columns from the builders into an arrow record.
-func (fb *flamegraphBuilder) NewRecord() (arrow.Record, error) {
-	cleanupArrs := make([]arrow.Array, 0, 26+(2*len(fb.builderLabelFields)))
+func (fb *flamegraphBuilder) prepareNewRecord() error {
+	// TODO: Do we want to clean up the builders too?
+	cleanupArrs := make([]releasable, 0, 10)
 	defer func() {
 		for _, arr := range cleanupArrs {
 			arr.Release()
 		}
 	}()
+
 	// We have manually tracked the total cumulative value.
 	// Now we set/overwrite the cumulative value for the root row (which is always the 0 row in our flame graphs).
 	fb.builderCumulative.Set(0, fb.cumulative)
 	fb.builderDiff.Set(0, fb.diff)
 
-	// We have manually tracked each row's children.
-	// So now we need to iterate over all rows in the record and append their children.
-	// We cannot do this while building the rows as we need to append the children while iterating over the rows.
-	for i := 0; i < fb.builderCumulative.Len(); i++ {
-		if i == 0 {
-			fb.builderChildren.Append(true)
-			for _, sampleLabelChildren := range fb.rootsRow {
-				for _, child := range sampleLabelChildren {
-					fb.builderChildrenValues.Append(uint32(child))
-				}
-			}
-			continue
-		}
-		if len(fb.children[i]) == 0 {
-			fb.builderChildren.AppendNull() // leaf
-		} else {
-			fb.builderChildren.Append(true)
-			for _, child := range fb.children[i] {
-				fb.builderChildrenValues.Append(uint32(child))
-			}
+	// At this point we don't care for the individual label roots anymore.
+	// We have grouped by them and their metadata before.
+	// We can combine all of them into the final children root
+	// making our lives easier later on when trimming and then creating the arrow array.
+	fb.children[0] = nil
+
+	// TODO: This shouldn't be here. Either the new hashing is stable or it should be fixed in the tests.
+	rootsRowKeys := make([]uint64, 0, len(fb.rootsRow))
+	for k := range fb.rootsRow {
+		rootsRowKeys = append(rootsRowKeys, k)
+	}
+	sort.Slice(rootsRowKeys, func(i, j int) bool { return rootsRowKeys[i] < rootsRowKeys[j] })
```

Comment:
This needs to be removed.
It works without this, but the tests won't be in a stable order anymore.

We need to fix the tests to sort the first level of rows for the group-by-pprof-labels sub test.

Reply:
Great thanks!